kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/row/BigQueryWriter.java

```diff
@@ -89,6 +90,24 @@ protected InsertAllRequest createInsertAllRequest(PartitionedTableId tableId,
         .build();
   }
 
+  /**
+   * Update rows to include only those failed in previous write.
+   * @param rows The rows to write.
+   * @param failedRows The set of indexes of row that failed to write.
+   */
+  protected void updateRowsBaseOnFailure(List<InsertAllRequest.RowToInsert> rows, Set<Long> failedRows) {
+    logger.debug("updating rows to include only failed rows during write operation");
+    ListIterator<InsertAllRequest.RowToInsert> iter = rows.listIterator(rows.size());
+    while (iter.hasPrevious()) {
+      if (!failedRows.contains((long)iter.previousIndex())) {
+        iter.previous();
+        iter.remove();
```

Comment:
Thinking about this more, although I don't see any problems right now with modifying `rows` underneath us, I am sort of worried it could cause unexpected problems down the line.

I like the idea of re-using the existing retry logic a lot, but I'm wondering if there is a good way to do it without in-place modifying our row list.